internal/user/ws.go

```diff
@@ -20,42 +22,87 @@ func (s *Service) InitSocketConnection(w http.ResponseWriter, r *http.Request, u
 	}
 	s.clientAdd(connection)
 	log.Printf("%s goes online.", u.Username)
+	s.logger.Entry.Infof("%s goes online.", u.Username)
+	err = s.storage.UpdateOnline(r.Context(), u, true)
+	if err != nil {
+		s.logger.Entry.Errorf("Error with update online status: %w", err)
+	}
+
+	cnt, err := s.storage.GetUnreadMessagesCount(r.Context(), u)
+	if err != nil {
+		s.logger.Entry.Errorf("Error with getting unread messages count: %w", err)
+	}
+
+	unreadMes, err := s.storage.GetUnreadMessages(r.Context(), u, cnt)
+	if err != nil {
+		s.logger.Entry.Errorf("Error with getting unread messages: %w", err)
+	}
+
+	for _, mes := range unreadMes {
+		err := connection.WriteJSON(mes)
+		if err != nil {
+			s.logger.Entry.Errorf("Error with sending mesage: %w", err)
+			err = connection.Close()
+			if err != nil {
+				s.logger.Entry.Errorf("error happens: %w", err)
+			}
+			s.clientDelete(connection)
+		}
+	}
+
 	if err = s.reader(connection, r.Context(), u); err != nil {
-		s.logger.Entry.Errorf("%s", err)
+		s.logger.Entry.Errorf("%w", err)
 	}
 	s.clientDelete(connection)
 	err = connection.Close()
 	if err != nil {
-		s.logger.Entry.Errorf("error happens: %s", err)
+		s.logger.Entry.Errorf("error happens: %w", err)
 	}
 	return nil
 }
 
-func (s *Service) reader(conn *websocket.Conn, ctx context.Context, u *User) (err error) {
+func (s *Service) reader(conn *websocket.Conn, ctx context.Context, u *models.User) (err error) {
 	for {
 		messageType, p, err := conn.ReadMessage()
 		if err != nil || messageType == websocket.CloseMessage {
 			err = s.storage.UpdateOnline(ctx, u, false)
 			log.Printf("<%s> left chat", u.Username)
-			s.logger.Entry.Errorf("Error with update online status: %s", err)
-			return fmt.Errorf("received error: %s", err)
-		}
-		err = s.storage.UpdateOnline(ctx, u, true)
-		if err != nil {
-			s.logger.Entry.Errorf("Error with update online status: %s", err)
+			s.logger.Entry.Errorf("Error with update online status: %w", err)
+			return fmt.Errorf("received error: %w", err)
 		}
 
 		s.rwMutex.RLock()
 		defer s.rwMutex.RUnlock()
+		var wg sync.WaitGroup
+
+		msg := &models.Message{
+			Text:  string(p),
+			UFrom: u.ID,
+		}
+		mid, err := s.storage.StoreMessage(ctx, msg)
+		if err != nil {
+			s.logger.Entry.Errorf("Error with store message: %s", err)
+		}
+
+		m, err := s.storage.FindOneMessage(ctx, mid)
+		if err != nil {
+			s.logger.Entry.Errorf("Error with getting message: %s", err)
+		}
+
 		for connKey := range s.clients {
-			if conn != connKey {
-				log.Printf("<%s>: %s", u.Username, string(p))
-				err := connKey.WriteMessage(messageType, p)
-				if err != nil {
-					s.logger.Entry.Errorf("Error with sending message: %s", err)
+			wg.Add(1)
+			go func(connKey *websocket.Conn, m *models.Message) {
```

Comment:
nit: It is not a good approach to run too many goroutines. You cannot control an amount of clients that is why you cannot control an amount of goroutines. If you want to send messages in parallel just run separate worker that listens a channel, and here message can be send in channel.